### PR TITLE
[Traceable FSDP2] [Inductor] Delete output buffer of inplace collective or wait_tensor after the op

### PR DESCRIPTION
**Pull request by yf225**
It should always be safe to attempt to `del` the output of inplace-collective / wait_tensor right after the op, because downstream should depend on the input (instead of the output) of the op. This is important for being able to release collective output memory as soon as possible (by decreasing the collective output tensor's refcount whenever possible).

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -8117,6 +8117,17 @@ def create_out_of_place(
             packed.outputs = [packed]
             return packed
 
+    def codegen(self, wrapper):
+        super().codegen(wrapper)
+        # NOTE: It should always be safe to attempt to `del` the output of inplace-collective / wait_tensor right after the op,
```
Comment (yifuwang) [inline torch/_inductor/ir.py:8122]: Just to refresh my memory - if we have:
```
buf_0 = ...
buf_1 = torch.ops._c10d_functional.all_reduce_(buf_0)
buf_2 = torch.ops._c10d_functional.wait_tensor(buf_0)?
```

Will the subsequent use of the all-reduce input always be `buf_0`

Comment (yf225) [reply torch/_inductor/ir.py:8122]: Yes it will - I updated the unit test `test_inductor_all_reduce_single` to show this

**Comment by yf225** [comment]
I ran into an interesting error on the AOTInductor test:

```cpp
auto buf2 = op__c10d_functional_all_reduce_coalesced__.call({buf0, buf1}, "avg", "0");
buf2.reset();
```
above is the generated AOTI C++ code. Here `buf2` is actually a `std::vector<at::Tensor>`, and `buf2.reset()` doesn't work out of the box. Instead we need to iterate through the vector and call reset on each individual tensor.

cc. @yifuwang @desertfire

File: torch/_inductor/ir.py (hunk 1 above)
Comment (desertfire) [reply torch/_inductor/ir.py:8089]: Is `_CollectiveKernel` always in-place? I am not familiar with collective ops, but I see this class contains both `create_inplace` and `create_out_of_place`.

File: torch/_inductor/codegen/cpp_wrapper_cpu.py [hunk 2]
```diff
@@ -1422,17 +1422,28 @@ def can_stack_allocate_buffer(self, buffer):
...
+        elif (
+            isinstance(buffer, ir.ExternKernel)
+            and isinstance(buffer.op_overload, torch._ops.OpOverload)
+            and isinstance(buffer.op_overload._schema.returns[0].type, torch.ListType)
+            and isinstance(
+                buffer.op_overload._schema.returns[0].type.getElementType(),
+                torch.TensorType,
+            )
+        ):
+            return f"for (auto t : {buffer.get_name()}) t.reset();"
```
Comment (desertfire) [inline torch/_inductor/codegen/cpp_wrapper_cpu.py:1444]: You can use reference (`auto& t`) here. Also it's clearer to wrap `t.reset();` with `{}`.

File: torch/_inductor/codegen/cpp_wrapper_cpu.py [hunk 3]
```diff
@@ -1422,17 +1422,28 @@ def can_stack_allocate_buffer(self, buffer):
...
                 and buffer.get_name() in V.graph.graph_inputs
             )
-            else f"{buffer.get_name()}.reset();"
-        )
+        ):
+            return ""
+        elif (
+            isinstance(buffer, ir.ExternKernel)
+            and isinstance(buffer.op_overload, torch._ops.OpOverload)
+            and isinstance(buffer.op_overload._schema.returns[0].type, torch.ListType)
```
Comment (desertfire) [inline torch/_inductor/codegen/cpp_wrapper_cpu.py:1438]: `returns[0]` works for your case, but maybe we should generalize the logic to all `returns`?

File: torch/_inductor/ir.py (hunk 1 above)
Comment (yf225) [reply torch/_inductor/ir.py:8123]: `_CollectiveKernel` can either be in-place or out-of-place. I just added a flag `self.inplace` to make it more clear whether a _CollectiveKernel is in-place.

File: torch/_inductor/codegen/cpp_wrapper_cpu.py (hunk 3 above)
Comment (yf225) [reply torch/_inductor/codegen/cpp_wrapper_cpu.py]: I believe `len(returns) > 1` is the MultiOutputLayout case, and based on existing code we explicitly don't generate buffer free code for it.

**Review by desertfire — APPROVED** [review]

**Review by yifuwang — APPROVED** [review]
Thanks this looks great!

**Comment by yifuwang** [comment]
@yf225 - @Chillee also also ran into this problem and proposed a more general fix for this: https://github.com/pytorch/pytorch/pull/129188. It fixes the issue for all `NonLayout` usage and not just for collective wait. Shall we go with the more general fix instead?

**Comment by Chillee** [comment]
imo I think the other PR is the right approach here

**Comment by yf225** [comment]
Closing in favor of https://github.com/pytorch/pytorch/pull/129188

**Pull request closed without merging**